### ISSUE TITLE
[v4] Custom field plugin migration journey

### ISSUE THREAD

**Issue by akcyp** [strapi/strapi#12506]
## Feature request, documentation update request and some questions

Hi, I have recently been migrating my custom field plugin (https://github.com/akcyp/strapi-plugin-point-list). Due to the little information on this subject, I have explored the topic from various sources and mostly by reading the source code.

## 1. Incorrect documentation / schema validation
Let's start with a bug in the documentation and an incorrect error message when installing the plugin
In https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/optional/plugins.html we have information that `resolve` parameter is optional but without it strapi couldn't load package (also in the error message it should say "resolve" instead of "path" I think).

![image](https://user-images.githubusercontent.com/67923777/153783602-19558b59-4199-4b55-b518-6007a73fff7e.png)

So, the only option is to create following config:
```js
module.exports = {
  'strapi-plugin-point-list': {
    enabled: true,
    resolve: 'node_modules/strapi-plugin-point-list'
  },
}
```
(strapi is looking for relative path that's why `node_modules` is needed)
Or maybe it is not a problem with documentation, but with config schema validation?

## 2. `fieldRenderer` attribute proposition
In version 4 the database system has been changed dramatically, and the `type` in content-type attributes does not mean exactly the same as in version 3. Previously I was using custom type and `columnType` (which is usless now because we only can modify predetermined types of data) to allow users use custom field component. Now changing `type` attribute is impossible because of static predetermined and non-extendable data types - there is no single file with all possible data types. Instead, we can have multiple files containing the same implementations - for example:
https://github.com/strapi/strapi/blob/master/packages/core/database/lib/fields.js
https://github.com/strapi/strapi/blob/master/packages/core/database/lib/types/index.js
https://github.com/strapi/strapi/blob/master/packages/plugins/graphql/server/services/constants.js
https://github.com/strapi/strapi/blob/master/packages/core/admin/admin/src/content-manager/components/Inputs/utils/getInputType.js

So we can assume that at the moment there is no point in changing the "type" attribute to your own. So why does the `GenericInput` (from `@stapi/helper-plugin`) component choose the component to display based on this property (and with `strapi.registerField` we can extend the list of allowed components)? If we should not change something on the server side, also on the client side it should not be allowed. So I propose to add an optional `fieldRenderer` attribute:

```diff
// src/api/example/content-type/schema.json
{
  "attributes": {
    "path": {
      "type": "string",
+     "fieldRenderer": "pointlist"
    }
  }
}
```

and changing https://github.com/strapi/strapi/blob/master/packages/core/helper-plugin/lib/src/components/GenericInput/index.js#L45
```js
const CustomInput = customInputs ? customInputs[rest.attribute.fieldRenderer || type] : null;
```

This is probably the easiest solution at the moment. In my package I do something similar with the help of overwriting (which is obviously not a very correct solution and can be troublesome) https://github.com/akcyp/strapi-plugin-point-list/blob/main/admin/src/index.js#L9-L20

## 3. Questions

- I saw an interesting and cool solution of `injectionZones` - will it be possible to add custom fields to the content-type builder in the future?
- How plugins should be versioned? Should the plugin version be equal to the strapi version?
- Is it possible to restore `multiple` prop to media library component? https://github.com/strapi/strapi/blob/master/packages/core/upload/admin/src/components/MediaLibraryDialog/index.js#L11
We can use it with the `useLibrary` hook, however the functionality of this component has become very limited in the new version

**Comment by smoothdvd** [comment]
@akcyp Yes, we hope Strapi will offer a way to implement custom input type.

BTW, I still can use columnType to create raw database type, something like:
```
{
  ...
  "attributes": {
    "effective_range": {
      "type": "string",
      "columnType": {
        "type": "specificType",
        "args": [
          "tstzrange"
        ]
      }
    }
  }
}
```

https://smoothdvd.medium.com/finally-found-the-correct-way-to-add-specific-type-column-in-strapi-v4-3b7f76e8e802

**Comment by dzcpy** [comment]
> @akcyp Yes, we hope Strapi will offer a way to implement custom input type.
> 
> BTW, I still can use columnType to create raw database type, something like:
> 
> ```
> {
>   ...
>   "attributes": {
>     "effective_range": {
>       "type": "string",
>       "columnType": {
>         "type": "specificType",
>         "args": [
>           "tstzrange"
>         ]
>       }
>     }
>   }
> }
> ```
> 
> https://smoothdvd.medium.com/finally-found-the-correct-way-to-add-specific-type-column-in-strapi-v4-3b7f76e8e802

I tried it, yes it can create a database field correctly, but I can't get it right saving it. There's always an error message. I somehow found a workaround as mentioned here:
https://github.com/akcyp/strapi-plugin-point-list/issues/1

**Comment by smoothdvd** [comment]
@dzcpy You should convert the Strapi type(like 'string') to raw data type in beforeCreate/Update lifecycle hooks.

**Comment by dzcpy** [comment]
> @dzcpy You should convert the Strapi type(like 'string') to raw data type in beforeCreate/Update lifecycle hooks.

Well, that's not really the issue. The problem is not that I can't save the data in a specific type. It's that in `v3` there is a workaround to customize the input field on content editing page in content manager. Now in `v4` it isn't working anymore. What I really like is to change the "type" so it can use my own component as field editing renderer, instead of the built-in one. It's not related to database, but a hard coded rule in the code which filters out all other types except those built-in ones.

**Comment by smoothdvd** [comment]
@dzcpy Sure. The only way to do this now is to modify the source code.

**Comment by dzcpy** [comment]
I'll paste my workaround for this feature in this thead, it works well for me, but isn't quite ideal if let's say a third party plugin also want to achive it:

1. Create a content type, add a `json` field (or `text` field in your case), specify a `componentType` (or whatever you'd like to call it):
```js
  "pluginOptions": {},
  "attributes": {
    "extra": {
      "type": "json",
      "componentType": "extra"
    }
  }
```

2. Create a plugin, during the `register` lifecycle ($pluginName/admin/src/index.js), overwrite the `json` field with your own component:
```js
    app.addFields([
      {
        type: 'json', Component: ExtraJSON
      },
    ]);
```

3.  In `ExtraJSON` component, check `props.attributes.componentType` field, if it's not `extra` then return the original `InputJSON` component, otherwise implement your own logic.

**Comment by timhonders** [comment]
I hope this gets fixed asap, we use custom inputs for a few customers but we can't upgrade to v4 du to this problem.

There is also a problem when adding a custom input to a component, you can't save because of hardcoded type validation.

**Comment by derrickmehaffy** [comment]
We are currently planning native custom field implementation but I'll let the expansion squad PM provide more details if they have any right now or in the future.

**Comment by rgesulfo** [comment]
Hi folks, 
The designer of the expansion squad at Strapi here 👋
I'm not super technical, so I won't be able to help as such, but feel free to ping me if you'd like to see a prototype of what we have in mind for custom fields :)
Thanks.

**Comment by timhonders** [comment]
@rgesulfo i like to see that prototype :)

**Comment by dzcpy** [comment]
Besides custom fields, I also hope to see that Strapi can ditch the custom design system and use some mature frameworks like `antd` or `mantine`

**Comment by derrickmehaffy** [comment]
> Besides custom fields, I also hope to see that Strapi can ditch the custom design system and use some mature frameworks like `antd` or `mantine`

@dzcpy won't happen right now. We spent a lot of time on our design system and we will be investing heavily in it for the future.

**Comment by moonlitusun** [comment]
> We are currently planning native custom field implementation but I'll let the expansion squad PM provide more details if they have any right now or in the future.

Please

**Comment by akcyp** [comment]
Hello, any update on this?

**Comment by derrickmehaffy** [comment]
Please see: https://feedback.strapi.io/developer-experience/p/custom-fields for more information about custom fields

**Comment by derrickmehaffy** [comment]
Moving this to the documentation team as I think most of the feedback here is around the docs for plugins
